swarm/util.py

```diff
@@ -0,0 +1,26 @@
+#  -*- coding: utf-8 -*-
+
+import functools
+
+
+def collector(aggregator_func):
+    """Combines the output of a generator with *aggregator_func*
+        >>> @collector(" ".join)
+        ... def string_yielder():
+        ...     yield "Hello"
+        ...     yield "World!"
+        "Hello World!
+    Allows the usage of yields in the code without needing to convert to list allowing
+    for more expressive and cleaner code instead of ``list.append``
+    This is particularly nice when interfacing with code that recieves data in a
+    paginated fashion but you'd like to return a ``pd.DataFrame`` or ``np.array``
+    This can also be chained.
+    """
+    def wrapper(func):
+        @functools.wraps(func)
+        def inner(*args, **kwargs):
+            return aggregator_func(func(*args, **kwargs))
+
+        return inner
+
+    return wrapper
```

Comment:
Ok, this looks 'artisinal as fuck' to use your own words.  Confirmed by the fact that there's longer docstrings than code. I think I like the motives behind it, maybe, because returning nice flat data that can be put into a df or array is great. But it sounds like you're a massive closure fan, and pushing the barriers beyond what most would.  I need to be persuaded!

Reply:
Haha, it's definitely artisanal as fuck. This is from a pull request I've made to `more_itertools` that has gone nowhere, but I wanted it anyway. It's definitely gratuitous